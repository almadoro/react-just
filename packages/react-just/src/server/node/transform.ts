import { IncomingMessage } from "node:http";
import { TLSSocket } from "node:tls";
import { Request } from "../../../types/server";

export function incomingMessageToRequest(
  incomingMessage: IncomingMessage,
): Request {
  const { method, headers: rawHeaders, url = "" } = incomingMessage;

  if (method !== "GET")
    throw new Error(
      `Method ${method} not supported. Only GET requests are supported.`,
    );

  const headers = new Headers();
  for (const [key, value] of Object.entries(rawHeaders)) {
    if (Array.isArray(value)) {
      for (const v of value) {
        headers.append(key, v);
      }
    } else if (typeof value === "string") {
      headers.set(key, value);
    }
  }

  const isHttps =
    incomingMessage.socket instanceof TLSSocket ||
    incomingMessage.headers["x-forwarded-proto"] === "https";

  const protocol = isHttps ? "https" : "http";

  const host =
    incomingMessage.headers["x-forwarded-host"] || headers.get("host");

  return {
    url: new URL(url, `${protocol}://${host}`),
    headers,
  };
}
