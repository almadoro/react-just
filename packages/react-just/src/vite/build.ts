import path from "node:path";
import type { Plugin } from "vite";
import { getRegisterModuleIdFromPath } from "./utils/client";

type BuildOptions = { entry: string };

export default function build(options: BuildOptions): Plugin {
  const clientEntries: string[] = [];

  return {
    name: "react-just:build",
    apply: "build",
    config() {
      return {
        appType: "custom",
        builder: {
          sharedConfigBuild: true,
          async buildApp(builder) {
            await builder.build(builder.environments.server);
            // It is required that the server build is executed first to catch all
            // the client entries.
            await builder.build(builder.environments.client);
          },
        },
        environments: {
          server: {
            consumer: "server",
            define: { "process.env.NODE_ENV": JSON.stringify("production") },
            build: {
              outDir: "dist/server",
              rollupOptions: {
                input: { index: path.resolve(options.entry) },
                output: { entryFileNames: "index.js" },
              },
            },
          },
          client: {
            consumer: "client",
            define: { "process.env.NODE_ENV": JSON.stringify("production") },
            build: {
              minify: false,
              outDir: "dist/client",
              rollupOptions: {
                input: { index: ENTRY_MODULE_ID },
                output: { entryFileNames: "index.js" },
              },
            },
          },
        },
      };
    },
    moduleParsed(moduleInfo) {
      if (this.environment.name !== "server") return;

      const isClientFile = moduleInfo.meta.reactUseClient?.transformed;
      if (isClientFile) clientEntries.push(moduleInfo.id);
    },
    resolveId(id) {
      if (id === ENTRY_MODULE_ID) return RESOLVED_ENTRY_MODULE_ID;
    },
    load(id) {
      if (id === RESOLVED_ENTRY_MODULE_ID)
        return getClientEntryCode(clientEntries);
    },
  };
}

export const ENTRY_MODULE_ID = "/virtual:client-entry";

const RESOLVED_ENTRY_MODULE_ID = "\0" + ENTRY_MODULE_ID;

function getClientEntryCode(entries: string[]) {
  let code = `import { hydrateFromWindowFlight, registerModule } from "react-just/client";`;

  for (let i = 0; i < entries.length; i++) {
    const entryPath = entries[i];
    const moduleId = getRegisterModuleIdFromPath(entryPath);
    code += `import * as entry_${i} from "${entryPath}";`;
    code += `registerModule("${moduleId}", entry_${i});`;
  }

  code += "hydrateFromWindowFlight();";
  return code;
}
