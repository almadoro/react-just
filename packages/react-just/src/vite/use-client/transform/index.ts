import { Program } from "estree";
import { traverse } from "estree-toolkit";
import transformExportDefaultDeclaration from "./export-default-declaration";
import transformExportNamedDeclaration from "./export-named-declaration";
import transformExportNamedFromSource from "./export-named-from-source";
import transformExportNamedSpecifiers from "./export-named-specifiers";
import Generator from "./generator";
import Module from "./module";

export default function transform(program: Program, options: TransformOptions) {
  const { generator, treeshakeImplementation } = options;

  const module = new Module(program);

  const useClientDirective = getUseClientDirective(program);

  if (!useClientDirective) return { transformed: false };

  module.remove(useClientDirective);

  module.unshift(generator.createRegisterFunctionImport());

  traverse(program, {
    ExportAllDeclaration() {
      throw new Error(
        "export all (`export *`) declarations are not supported on client modules",
      );
    },
    ExportNamedDeclaration(path) {
      const node = path.node!;
      if (node.source) {
        transformExportNamedFromSource(node, module, generator);
      } else if (node.declaration) {
        transformExportNamedDeclaration(node, module, generator);
      } else {
        transformExportNamedSpecifiers(node, module, generator);
      }
    },
    ExportDefaultDeclaration(path) {
      const node = path.node!;
      transformExportDefaultDeclaration(node, module, generator);
    },
  });

  if (treeshakeImplementation)
    program.body = generator.createTreeshakedBody(program.body);

  return { transformed: true };
}

export type TransformOptions = {
  generator: Generator;
  treeshakeImplementation: boolean;
};

const USE_CLIENT_DIRECTIVE = "use client";

function getUseClientDirective(program: Program) {
  // The "use client" directive must be the first node in the file.
  const firstNode = program.body[0];

  if (
    firstNode &&
    firstNode.type === "ExpressionStatement" &&
    firstNode.expression.type === "Literal" &&
    firstNode.expression.value === USE_CLIENT_DIRECTIVE
  )
    return firstNode;

  return null;
}
