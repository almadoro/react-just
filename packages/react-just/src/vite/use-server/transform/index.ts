import {
  ArrowFunctionExpression,
  FunctionDeclaration,
  FunctionExpression,
  Program,
} from "estree";
import { NodePath, traverse } from "estree-toolkit";
import {
  getIsUseServerDirective,
  getUseServerModuleDirective,
} from "../directive";
import transformExportDefaultDeclaration from "./export-default-declaration";
import transformExportNamedDeclaration from "./export-named-declaration";
import transformExportNamedFromSource from "./export-named-from-source";
import transformExportNamedSpecifiers from "./export-named-specifiers";
import Generator from "./generator";
import Module from "./module";

export default function transform(program: Program, options: TransformOptions) {
  const { generator, treeshakeImplementation } = options;

  const module = new Module(program);

  // The directive "use server" directive can appear at the top
  // of a module.
  const useServerModuleDirective = getUseServerModuleDirective(program);

  if (useServerModuleDirective) module.remove(useServerModuleDirective);

  traverse(program, {
    ExportAllDeclaration() {
      if (!useServerModuleDirective) return;

      throw new Error(
        'export all (`export *`) declarations are not supported on "use server" modules',
      );
    },
    ExportNamedDeclaration(path) {
      if (!useServerModuleDirective) return;

      const node = path.node!;
      if (node.source) {
        transformExportNamedFromSource(node, module, generator);
      } else if (node.declaration) {
        transformExportNamedDeclaration(node, module, generator);
      } else {
        transformExportNamedSpecifiers(node, module, generator);
      }
    },
    ExportDefaultDeclaration(path) {
      if (!useServerModuleDirective) return;

      const node = path.node!;
      transformExportDefaultDeclaration(node, module, generator);
    },
    // The directive "use server" directive can appear at the top of a function.
    ArrowFunctionExpression: FunctionLike,
    FunctionDeclaration: FunctionLike,
    FunctionExpression: FunctionLike,
  });

  if (!useServerModuleDirective)
    throw new Error('Expected "use server" directive to exist');

  module.unshift(generator.createRegisterFunctionImport());

  if (treeshakeImplementation)
    program.body = generator.createTreeshakedBody(program.body);
}

export type TransformOptions = {
  generator: Generator;
  treeshakeImplementation: boolean;
};

function FunctionLike(
  path: NodePath<
    FunctionDeclaration | FunctionExpression | ArrowFunctionExpression
  >,
) {
  const node = path.node!;

  if (node.body.type !== "BlockStatement") return;

  const firstBlockNode = node.body.body[0];

  const isUseServerDirective =
    firstBlockNode && getIsUseServerDirective(firstBlockNode);

  if (!isUseServerDirective) return;

  const async = node.async;

  if (!async) throw new Error("server functions must be async");

  throw new Error("use server directive inside functions not supported yet");
}
