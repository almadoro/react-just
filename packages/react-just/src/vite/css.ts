import { DevEnvironment, Plugin } from "vite";
import { isFlightEnvironment } from "./environments";
import { isCssModule } from "./utils";

export default function css(): Plugin {
  const flightEnvironments: DevEnvironment[] = [];

  return {
    name: "react-just:css",
    configureServer(server) {
      for (const environment of Object.values(server.environments)) {
        if (isFlightEnvironment(environment.name))
          flightEnvironments.push(environment);
      }
    },
    resolveId(id) {
      if (id === CSS_MODULES) return RESOLVED_CSS_MODULES;
    },
    load(id) {
      if (id === RESOLVED_CSS_MODULES)
        return getCssModulesCode(flightEnvironments);
    },
  };
}

export const CSS_MODULES = "/virtual:react-just/css-modules";
export const RESOLVED_CSS_MODULES = "\0" + CSS_MODULES;

function getCssModulesCode(flightEnvironments: DevEnvironment[]) {
  const ids = new Set<string>();

  for (const environment of flightEnvironments) {
    const { idToModuleMap } = environment.moduleGraph;

    for (const [id, module] of idToModuleMap.entries()) {
      if (
        isCssModule(id) &&
        // On development, map may contain unused modules. Ignore them.
        module.importers.size !== 0
      )
        ids.add(id);
    }
  }

  return Array.from(ids)
    .map((id) => `import "${id}";`)
    .join("\n");
}
