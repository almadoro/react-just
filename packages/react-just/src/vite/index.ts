import react from "@vitejs/plugin-react";
import type { PluginOption } from "vite";
import build from "./build";
import dev from "./dev";
import { ReactJustOptions } from "./types";
import useClient from "./use-client";
import useServer from "./use-server";

const FLIGHT_MIME_TYPE = "text/x-component";

export default function reactJust(options: ReactJustOptions): PluginOption {
  return [
    build({ app: options.app, flightMimeType: FLIGHT_MIME_TYPE }),
    dev({ app: options.app, flightMimeType: FLIGHT_MIME_TYPE }),
    react(),
    useClient(),
    useServer(),
  ];
}
