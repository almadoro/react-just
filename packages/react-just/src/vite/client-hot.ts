import { DevEnvironment, Plugin } from "vite";
import { CSS_MODULES } from "./css";
import { ENVIRONMENTS, isFlightEnvironment } from "./environments";
import { CLIENT_MODULES } from "./use-client";

export default function clientHot(): Plugin {
  return {
    name: "react-just:client-hot",
    configureServer(server) {
      watchClientReactChange(server.environments[ENVIRONMENTS.CLIENT]);
    },
    hotUpdate(ctx) {
      if (isFlightEnvironment(this.environment.name))
        onFlightHotUpdate(ctx.server.environments[ENVIRONMENTS.CLIENT]);
    },
    resolveId(id) {
      switch (id) {
        case CLIENT_HOT_MODULES:
          return RESOLVED_CLIENT_HOT_MODULES;
        case BROWSER_HASH_CHECK:
          return RESOLVED_BROWSER_HASH_CHECK;
        case HMR_PREAMBLE:
          return RESOLVED_HMR_PREAMBLE;
        case SERVER_HMR:
          return RESOLVED_SERVER_HMR;
      }
    },
    load(id) {
      switch (id) {
        case RESOLVED_CLIENT_HOT_MODULES:
          return getClientHotModulesCode();
        case RESOLVED_BROWSER_HASH_CHECK:
          if (!(this.environment instanceof DevEnvironment)) return "";
          const browserHash =
            this.environment.depsOptimizer?.metadata.browserHash;
          if (!browserHash)
            throw new Error("Expected browser hash to be available");
          return getBrowserHashCheckCode(browserHash);
        case RESOLVED_HMR_PREAMBLE:
          return getHmrPreambleCode();
        case RESOLVED_SERVER_HMR:
          return getServerHmrCode();
      }
    },
  };
}

const BROWSER_HASH_CHECK_HOT_EVENT = "react-just:browser-hash-check";

function watchClientReactChange(clientEnv: DevEnvironment) {
  let currentBrowserHash = clientEnv.depsOptimizer?.metadata.browserHash;

  // When a reoptimization occurs, the React version of loaded modules on
  // client will not match future loaded ones. A full reload is required.
  setInterval(() => {
    let prevBrowserHash = currentBrowserHash;
    currentBrowserHash = clientEnv.depsOptimizer?.metadata.browserHash;

    const needsReload = prevBrowserHash !== currentBrowserHash;
    if (needsReload) clientEnv.hot.send({ type: "full-reload" });
  }, 0);

  // A reoptimization may occur before the Vite connection is established;
  // thus, any triggered full-reload events won't be received. A manual check
  // is required after the connection is established.
  clientEnv.hot.on("connection", () => {
    // The event is sent to every connected client. This doesn't matter.
    // This event should have no effects when the browser's modules are up to
    // date.
    clientEnv.hot.send({
      type: "custom",
      event: BROWSER_HASH_CHECK_HOT_EVENT,
      data: { browserHash: currentBrowserHash },
    });
  });
}

const TREE_RELOAD_HOT_EVENT = "react-just:tree-reload";

function onFlightHotUpdate(clientEnv: DevEnvironment) {
  // Send an event to trigger the client modules and react tree reload.
  clientEnv.hot.send({
    type: "custom",
    event: TREE_RELOAD_HOT_EVENT,
    data: { eventId: crypto.randomUUID() },
  });
}

export const CLIENT_HOT_MODULES = "/virtual:react-just/client-hot-modules";
const RESOLVED_CLIENT_HOT_MODULES = "\0" + CLIENT_HOT_MODULES;

function getClientHotModulesCode() {
  return (
    `import "${BROWSER_HASH_CHECK}";` +
    `import "${HMR_PREAMBLE}";` +
    `import "${SERVER_HMR}";`
  );
}

const BROWSER_HASH_CHECK = "/virtual:react-just/browser-hash-check";
const RESOLVED_BROWSER_HASH_CHECK = "\0" + BROWSER_HASH_CHECK;

function getBrowserHashCheckCode(currentBrowserHash: string) {
  return (
    `if (import.meta.hot) {` +
    `  import.meta.hot.on("${BROWSER_HASH_CHECK_HOT_EVENT}", ({ browserHash }) => {` +
    `    if (browserHash !== "${currentBrowserHash}")` +
    `      window.location.reload();` +
    `  });` +
    `}`
  );
}

const HMR_PREAMBLE = "/virtual:react-just/hmr-preamble";
const RESOLVED_HMR_PREAMBLE = "\0" + HMR_PREAMBLE;

function getHmrPreambleCode() {
  // Taken from: https://github.com/vitejs/vite-plugin-react/blob/main/packages/common/refresh-utils.ts
  return (
    `import { injectIntoGlobalHook } from "/@react-refresh";` +
    `if (import.meta.hot) {` +
    `  injectIntoGlobalHook(window);` +
    `  window.$RefreshReg$ = () => {};` +
    `  window.$RefreshSig$ = () => (type) => type;` +
    `}`
  );
}

const SERVER_HMR = "/virtual:react-just/server-hmr";
const RESOLVED_SERVER_HMR = "\0" + SERVER_HMR;

function getServerHmrCode() {
  return (
    `import { createFromRscFetch, render, RSC_MIME_TYPE } from "react-just/client";` +
    `if (import.meta.hot) {` +
    `  let lastEventId = null;` +
    `  import.meta.hot.on("${TREE_RELOAD_HOT_EVENT}", ({ eventId }) => {` +
    `    lastEventId = eventId;` +
    `    const headers = { accept: RSC_MIME_TYPE };` +
    // Don't use exactly the same URL as the one we're trying to load to avoid
    // sharing cache between the RSC and the HTML.
    `    const url = new URL(window.location.href);` +
    `    url.searchParams.set("__rsc__", "1");` +
    `    createFromRscFetch(fetch(url, { headers })).then(async (tree) => {` +
    // Add a timestamp to trigger modules reload on the browser.
    `      await import(/* @vite-ignore */ "${CLIENT_MODULES}?t=" + Date.now());` +
    `      await import(/* @vite-ignore */ "${CSS_MODULES}?t=" + Date.now());` +
    // Avoid race conditions between multiple reloads. Render only the latest one.
    `      if (lastEventId === eventId) render(tree);` +
    `    });` +
    `  });` +
    `}`
  );
}
