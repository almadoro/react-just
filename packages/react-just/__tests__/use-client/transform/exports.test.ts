import transform, { TransformOptions } from "@/vite/use-client/transform";
import Generator from "@/vite/use-client/transform/generator";
import { generate } from "astring";
import { builders } from "estree-toolkit";
import fs from "node:fs/promises";
import path from "node:path";
import { parseAstAsync } from "vite";
import { describe, expect, test } from "vitest";

const MODULE_ID = "module.js";

// client-like environment
const noTreeshakeEnv: TransformOptions = {
  generator: new Generator({
    getRegisterArguments: ({ exportName, implementationIdentifier }) => [
      builders.identifier(implementationIdentifier),
      builders.literal(MODULE_ID),
      builders.literal(exportName),
    ],
    registerClientReferenceSource: "react-just/no-treeshake",
  }),
  treeshakeImplementation: false,
};

// flight-like environment
const treeshakeEnv: TransformOptions = {
  generator: new Generator({
    getRegisterArguments: ({ exportName }) => [
      builders.literal(MODULE_ID),
      builders.literal(exportName),
    ],
    registerClientReferenceSource: "react-just/treeshake",
  }),
  treeshakeImplementation: true,
};

describe(
  "transforms exports on a no-treeshake environment",
  exportsTests(noTreeshakeEnv),
);
describe(
  "transforms exports on a treeshake environment",
  exportsTests(treeshakeEnv),
);

function exportsTests(options: TransformOptions) {
  return () => {
    describe("export default declarations", () => {
      test(
        "class declarations",
        transformTest("export-default-declaration/class-declaration.js"),
      );

      test(
        "function declarations",
        transformTest("export-default-declaration/function-declaration.js"),
      );

      test(
        "identifier",
        transformTest("export-default-declaration/identifier.js"),
      );

      test(
        "value declarations",
        transformTest("export-default-declaration/value-declaration.js"),
      );
    });

    describe("export named declarations", () => {
      test(
        "array pattern",
        transformTest("export-named-declaration/array-pattern.js"),
      );

      test(
        "assignment pattern",
        transformTest("export-named-declaration/assignment-pattern.js"),
      );

      test("class", transformTest("export-named-declaration/class.js"));

      test(
        "identifier",
        transformTest("export-named-declaration/identifier.js"),
      );

      test("function", transformTest("export-named-declaration/function.js"));

      test(
        "object pattern",
        transformTest("export-named-declaration/object-pattern.js"),
      );

      test(
        "rest element",
        transformTest("export-named-declaration/rest-element.js"),
      );
    });

    test(
      "export named from source",
      transformTest("export-named-from-source.js"),
    );

    test(
      "export named specifiers",
      transformTest("export-named-specifiers.js"),
    );

    test("throws when there is an export all declaration", async () => {
      const code = "'use client'; export * from 'pkg'";

      const program = await parseAstAsync(code);

      expect(() => transform(program, options)).toThrowError();
    });
  };

  function transformTest(filepath: string) {
    return async () => {
      const code = await fs.readFile(
        path.resolve(import.meta.dirname, "fixtures", filepath),
        "utf-8",
      );

      const program = await parseAstAsync(code);

      transform(program, options);

      const output = generate(program);

      expect(output).toMatchSnapshot();
    };
  }
}
