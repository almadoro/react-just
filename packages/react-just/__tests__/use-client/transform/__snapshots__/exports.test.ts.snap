// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transforms exports on a no-treeshake environment > export default declarations > class declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
class A {}
const $$Ref$$default = $$registerClientReference$$(A, "module.js", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a no-treeshake environment > export default declarations > function declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
function $$default$$() {}
const $$Ref$$default = $$registerClientReference$$($$default$$, "module.js", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a no-treeshake environment > export default declarations > identifier 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
let a;
const $$Ref$$default = $$registerClientReference$$(a, "module.js", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a no-treeshake environment > export default declarations > value declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
const $$default$$ = 1;
const $$Ref$$default = $$registerClientReference$$($$default$$, "module.js", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > array pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
let [a, {b}, [c], d = 1, , ...e] = [];
const $$Ref$$a = $$registerClientReference$$(a, "module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$(b, "module.js", "b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerClientReference$$(c, "module.js", "c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerClientReference$$(d, "module.js", "d");
export {$$Ref$$d as d};
const $$Ref$$e = $$registerClientReference$$(e, "module.js", "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > assignment pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
let [a = 1, {b} = c, [d] = e] = [];
const $$Ref$$a = $$registerClientReference$$(a, "module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$(b, "module.js", "b");
export {$$Ref$$b as b};
const $$Ref$$d = $$registerClientReference$$(d, "module.js", "d");
export {$$Ref$$d as d};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > class 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
class A {}
const $$Ref$$A = $$registerClientReference$$(A, "module.js", "A");
export {$$Ref$$A as A};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > function 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
function a() {}
const $$Ref$$a = $$registerClientReference$$(a, "module.js", "a");
export {$$Ref$$a as a};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > identifier 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
let a;
const $$Ref$$a = $$registerClientReference$$(a, "module.js", "a");
export {$$Ref$$a as a};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > object pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
let {a, b: {c}, d: [e], f = 1, ...g} = {};
const $$Ref$$a = $$registerClientReference$$(a, "module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$(c, "module.js", "c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerClientReference$$(e, "module.js", "e");
export {$$Ref$$e as e};
const $$Ref$$f = $$registerClientReference$$(f, "module.js", "f");
export {$$Ref$$f as f};
const $$Ref$$g = $$registerClientReference$$(g, "module.js", "g");
export {$$Ref$$g as g};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > rest element 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
let [...a] = [];
let [...[b, c]] = [];
let [...{d: e}] = [];
const $$Ref$$a = $$registerClientReference$$(a, "module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$(b, "module.js", "b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerClientReference$$(c, "module.js", "c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerClientReference$$(e, "module.js", "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a no-treeshake environment > export named from source 1`] = `
"import {a as $$a$$, b as $$c$$, b as $$d$$, default as $$default$$, default as $$e$$} from "pkg";
import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
let c = 1;
const $$Ref$$a = $$registerClientReference$$($$a$$, "module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$($$c$$, "module.js", "c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerClientReference$$($$d$$, "module.js", "d");
export {$$Ref$$d as d};
const $$Ref$$default = $$registerClientReference$$($$default$$, "module.js", "default");
export {$$Ref$$default as default};
const $$Ref$$e = $$registerClientReference$$($$e$$, "module.js", "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a no-treeshake environment > export named specifiers 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/no-treeshake";
let a;
let b;
let d;
const $$Ref$$a = $$registerClientReference$$(a, "module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$(b, "module.js", "c");
export {$$Ref$$c as c};
const $$Ref$$default = $$registerClientReference$$(d, "module.js", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a treeshake environment > export default declarations > class declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$default = $$registerClientReference$$("module.js", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a treeshake environment > export default declarations > function declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$default = $$registerClientReference$$("module.js", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a treeshake environment > export default declarations > identifier 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$default = $$registerClientReference$$("module.js", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a treeshake environment > export default declarations > value declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$default = $$registerClientReference$$("module.js", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > array pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerClientReference$$("module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$("module.js", "b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerClientReference$$("module.js", "c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerClientReference$$("module.js", "d");
export {$$Ref$$d as d};
const $$Ref$$e = $$registerClientReference$$("module.js", "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > assignment pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerClientReference$$("module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$("module.js", "b");
export {$$Ref$$b as b};
const $$Ref$$d = $$registerClientReference$$("module.js", "d");
export {$$Ref$$d as d};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > class 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$A = $$registerClientReference$$("module.js", "A");
export {$$Ref$$A as A};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > function 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerClientReference$$("module.js", "a");
export {$$Ref$$a as a};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > identifier 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerClientReference$$("module.js", "a");
export {$$Ref$$a as a};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > object pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerClientReference$$("module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$("module.js", "c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerClientReference$$("module.js", "e");
export {$$Ref$$e as e};
const $$Ref$$f = $$registerClientReference$$("module.js", "f");
export {$$Ref$$f as f};
const $$Ref$$g = $$registerClientReference$$("module.js", "g");
export {$$Ref$$g as g};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > rest element 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerClientReference$$("module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$("module.js", "b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerClientReference$$("module.js", "c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerClientReference$$("module.js", "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a treeshake environment > export named from source 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerClientReference$$("module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$("module.js", "c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerClientReference$$("module.js", "d");
export {$$Ref$$d as d};
const $$Ref$$default = $$registerClientReference$$("module.js", "default");
export {$$Ref$$default as default};
const $$Ref$$e = $$registerClientReference$$("module.js", "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a treeshake environment > export named specifiers 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerClientReference$$("module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$("module.js", "c");
export {$$Ref$$c as c};
const $$Ref$$default = $$registerClientReference$$("module.js", "default");
export {$$Ref$$default as default};
"
`;
